include/deal.II/matrix_free/operators.h

```diff
@@ -717,6 +757,80 @@ namespace MatrixFreeOperators
     dst*= omega;
   }
 
+
+
+  //-----------------------------MassOperator----------------------------------
+
+  template <int dim, int fe_degree, int n_components, typename number>
+  MassOperator<dim, fe_degree, n_components, number>::
+  MassOperator ()
+    :
+    Base<dim, number>()
+  {}
+
+
+
+  template <int dim, int fe_degree, int n_components, typename number>
+  void
+  MassOperator<dim, fe_degree, n_components, number>::
+  compute_diagonal()
+  {
+    Assert((Base<dim, number>::data != NULL), ExcNotInitialized());
+
+    LinearAlgebra::distributed::Vector<number> ones;
+    Base<dim, number>::initialize_dof_vector(Base<dim, number>::inverse_diagonal_entries);
+    Base<dim, number>::initialize_dof_vector(ones);
+    Base<dim, number>::inverse_diagonal_entries = 1.;
+    ones = 1.;
+    ones.update_ghost_values();
+    apply_add(Base<dim, number>::inverse_diagonal_entries, ones);
+
+    const unsigned int local_size = Base<dim, number>::inverse_diagonal_entries.local_size();
+    for (unsigned int i=0; i<local_size; ++i)
+      {
+        Base<dim, number>::inverse_diagonal_entries.local_element(i)
```

Comment:
No, I set all elements to one beforehand and the constrained ones are not touched during `vmult_add`.